gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java

```diff
@@ -1334,47 +1354,32 @@ public void putAssignmentsAndCategoryItemsInMatrix(Map<String, GbStudentGradeInf
 
 						boolean gradeable = false;
 
-						for (final PermissionDefinition permission : permissions) {
-							// we know they are all GRADE so no need to check here
-
-							boolean categoryOk = false;
-							boolean groupOk = false;
+						//If TA does not have all the permissions over categories and groups
+						if (!allPermissions) {
 
-							final Long permissionCategoryId = permission.getCategoryId();
-							final String permissionGroupReference = permission.getGroupReference();
+							//Check category and group permissions
+							List<String> arr = permByCat.get(gradeCategoryId) != null ? permByCat.get(gradeCategoryId) : new ArrayList<String>();
 
-							log.debug("permissionCategoryId: {}", permissionCategoryId);
-							log.debug("permissionGroupReference: {}", permissionGroupReference);
-
-							// if permissions category is null (can grade all categories) or they match (can grade this category)
-							if (!categoriesAreEnabled() || (permissionCategoryId == null || permissionCategoryId.equals(gradeCategoryId))) {
-								categoryOk = true;
-								log.debug("Category check passed");
+							//add groups with permission in all categories
+							if (permByCat.containsKey(Long.valueOf(-1L))) {
```

Comment:
You're doing two map searches here  when you don't need to. You could just go:
arr.addAll(permByCat.getOrDefault(Long.valueOf(-1L), Collections.\<String\>emptyList());

or ...

Optional.ofNullable(permByCat.get(Long.valueOf(-1L))).ifPresent(arr::addAll)

from https://stackoverflow.com/questions/30739105/any-null-safe-alternative-to-arraylist-addall

I like the second one, tbh.